File: add_groceries.py
#!/usr/bin/env python3
"""
Picnic Grocery Automation Script (HTTP API Approach)
Uses direct HTTP requests instead of browser automation to avoid anti-bot detection
"""

import os
import sys
import time
import requests
import json
from typing import Optional, List

# Disable SSL warnings for cleaner output
requests.packages.urllib3.disable_warnings()

class PicnicAPI:
    def __init__(self, username: str, password: str):
        self.username = username
        self.password = password
        self.session = requests.Session()
        self.token = None

        # Realistic headers that look like a real mobile/web client
        self.session.headers.update({
            'User-Agent': 'Mozilla/5.0 (iPhone; CPU iPhone OS 16_0 like Mac OS X) AppleWebKit/605.1.15',
            'Accept': 'application/json, text/plain, */*',
            'Accept-Language': 'en-US,en;q=0.9',
            'Content-Type': 'application/json',
            'Origin': 'https://www.picnic.app',
            'Referer': 'https://www.picnic.app/',
        })

        self.added_items = []
        self.failed_items = []

    def login(self) -> bool:
        """Authenticate with Picnic API"""
        try:
            print("🔐 Authenticating with Picnic API...")

            url = "https://api.picnic.app/user/login"
            data = {
                "username": self.username,
                "password": self.password
            }

            response = self.session.post(url, json=data, verify=False)

            if response.status_code == 200:
                result = response.json()
                self.token = result.get('access_token')
                self.session.headers['Authorization'] = f'Bearer {self.token}'
                print("✅ Authentication successful!\n")
                return True
            else:
                print(f"❌ Login failed: {response.status_code}")
                print(response.text)
                return False

        except Exception as e:
            print(f"❌ Login error: {e}")
            return False

    def search_product(self, query: str) -> Optional[dict]:
        """Search for a product"""
        try:
            url = "https://api.picnic.app/search"
            params = {"q": query}

            response = self.session.get(url, params=params, verify=False, timeout=10)

            if response.status_code == 200:
                results = response.json()
                if results and len(results) > 0:
                    return results[0]
            return None

        except Exception as e:
            return None

    def add_to_cart(self, product_id: str, quantity: int = 1) -> bool:
        """Add product to cart"""
        try:
            url = f"https://api.picnic.app/cart/add_product"
            data = {
                "product_id": product_id,
                "quantity": quantity
            }

            response = self.session.post(url, json=data, verify=False, timeout=10)
            return response.status_code in [200, 201]

        except Exception as e:
            return False

    def search_and_add(self, item_name: str) -> bool:
        """Search for item and add to cart"""
        try:
            print(f"  Searching for: {item_name}...", end=" ", flush=True)

            product = self.search_product(item_name)

            if not product:
                print("❌ Not found")
                self.failed_items.append(item_name)
                return False

            product_id = product.get('id')
            if self.add_to_cart(product_id):
                print("✅ Added")
                self.added_items.append(item_name)
                return True
            else:
                print("❌ Failed to add")
                self.failed_items.append(item_name)
                return False

        except Exception as e:
            print(f"❌ Error")
            self.failed_items.append(item_name)
            return False

    def add_all_groceries(self, items: List[str]) -> None:
        """Add all items to cart"""
        print(f"📦 Adding {len(items)} items to cart...\n")

        for item in items:
            self.search_and_add(item)
            time.sleep(0.5)

    def show_summary(self) -> None:
        """Show what was added"""
        print("\n" + "=" * 50)
        print(f"✅ Added: {len(self.added_items)} items")
        if self.failed_items:
            print(f"❌ Failed: {len(self.failed_items)} items")
            print("\nFailed items (add manually):")
            for item in self.failed_items:
                print(f"  - {item}")

        print("\n📝 Next steps:")
        print("  1. Open Picnic app/website")
        print("  2. Review your cart")
        print("  3. Select delivery slot and checkout")
        print("\n✨ Done!")


# Grocery list
GROCERY_LIST = [
    "Sirtakis Joghurt nach griechischer Art 1kg",
    "Arla Skyr Natur 450g",
    "Cheddar",
    "Eier 18",
    "Haferdrink",
    "REWE Bio pflanzlich Barista Hafer-Soja Drink",
    "Butter",
    "Crème fraîche",
    "Kuhmilch",
    "Tomaten",
    "Gurke",
    "Kartoffeln",
    "Nüsse",
    "Äpfel",
    "Reiswaffeln",
    "Schär Meisterbäckers Vital glutenfrei",
    "Zitrone",
    "Delamaris Sardinen",
    "Heinz Ketchup ohne Zucker",
    "KoRo Erdnussbutter",
    "Schär Mix It Universal glutenfrei",
    "Zucchini",
    "Paprika",
    "Kirschtomaten",
    "Spinat",
    "Lauch",
    "Aubergine",
    "Galbani Mozzarella di Bufala",
    "Parmesan",
    "Rapunzel Bio gemahlene Mandeln",
    "Passierte Tomaten",
    "Basilikum",
    "Zwiebeln",
    "Knoblauch",
]


def main():
    username = os.getenv("PICNIC_USERNAME")
    password = os.getenv("PICNIC_PASSWORD")

    if not username:
        username = input("Enter your Picnic email: ")
    if not password:
        password = input("Enter your Picnic password: ")

    try:
        api = PicnicAPI(username, password)

        if not api.login():
            sys.exit(1)

        api.add_all_groceries(GROCERY_LIST)
        api.show_summary()

    except KeyboardInterrupt:
        print("\n\n⚠️  Cancelled by user")
    except Exception as e:
        print(f"\n❌ Error: {e}")


if __name__ == "__main__":
    main()

File: add_groceries.js
#!/usr/bin/env node

/**
 * Picnic Grocery Automation Script (Node.js)
 * Uses the official picnic-api package to automate adding items to cart
 */

const PicnicClient = require("picnic-api");
const readline = require("readline");
const { smartProductSearch } = require("./picnic-search.js");

// Grocery list - will be fetched from server if available
let GROCERY_LIST = [];

class PicnicAutomation {
  constructor(username, password) {
    this.username = username;
    this.password = password;
    this.client = new PicnicClient({ countryCode: "DE" });
    this.addedItems = [];
    this.failedItems = [];
    this.purchaseHistory = new Map(); // productName -> product object
  }

  async loadPurchaseHistory() {
    try {
      console.log("📜 Loading your purchase history...");
      // Get user's deliveries
      const deliveries = await this.client.delivery.getDeliveries();

      if (deliveries && deliveries.length > 0) {
        // Extract products from recent deliveries
        for (const delivery of deliveries.slice(0, 5)) {
          // Last 5 deliveries
          if (delivery.orders) {
            for (const order of delivery.orders) {
              if (order.products) {
                for (const product of order.products) {
                  const key = (product.name || "").toLowerCase();
                  this.purchaseHistory.set(key, product);
                }
              }
            }
          }
        }
        console.log(
          `✅ Loaded ${this.purchaseHistory.size} products from history\n`
        );
      }
    } catch (error) {
      console.log("⚠️  Could not load purchase history (continuing anyway)\n");
    }
  }

  async clearCart() {
    try {
      console.log("🗑️  Clearing cart...");
      await this.client.cart.clearCart();
      console.log("✅ Cart cleared\n");
      return true;
    } catch (error) {
      console.log("⚠️  Could not clear cart (continuing anyway)\n");
      return false;
    }
  }

  async login() {
    try {
      console.log("🔐 Logging in to Picnic...");
      const loginResponse = await this.client.auth.login(this.username, this.password);

      // Check if 2FA is required
      if (loginResponse.second_factor_authentication_required) {
        console.log("📱 2FA required. Requesting code via SMS...");

        try {
          // Generate 2FA code (send SMS)
          await this.client.auth.generate2FACode("SMS");
          console.log("✅ SMS sent! Check your phone for the code.\n");

          // Ask user for the code
          const code = await askQuestion("Enter your 6-digit SMS code: ");

          // Verify the code
          await this.client.auth.verify2FACode(code);
          console.log("✅ 2FA verified! Logged in.\n");
          return true;
        } catch (verifyError) {
          console.error("❌ 2FA failed:", verifyError.message);
          return false;
        }
      } else {
        console.log("✅ Login successful!\n");
        return true;
      }
    } catch (error) {
      console.error("❌ Login failed:", error.message);
      return false;
    }
  }

  selectBestProduct(results, itemName) {
    if (!results || results.length === 0) return null;

    // Score products by criteria
    let scored = results.map((product) => {
      let score = 0;
      const name = (product.name || "").toLowerCase();
      const brand = (product.brand || "").toLowerCase();

      // HIGHEST PRIORITY: Previously purchased
      if (this.purchaseHistory.has(name)) {
        score += 100; // Massive boost for items you've bought before
      }

      // Prefer Bio/organic
      if (name.includes("bio") || brand.includes("bio")) score += 10;

      // Prefer local/German brands
      if (
        brand.includes("rewe") ||
        brand.includes("rapunzel") ||
        brand.includes("schär")
      )
        score += 8;

      // Prefer items without artificial additives
      if (name.includes("natur") || name.includes("rein")) score += 5;

      // Price preference (lower is better, but not the only factor)
      if (product.price) score += Math.max(0, 10 - product.price / 10);

      return { product, score };
    });

    // Sort by score (highest first)
    scored.sort((a, b) => b.score - a.score);
    return scored[0].product;
  }

  async searchAndAddItem(itemName) {
    try {
      process.stdout.write(`  Searching for: ${itemName}... `);

      // Try exact search, then a purchase-history fuzzy match, then
      // progressively simplified/normalized fallback queries.
      const match = await smartProductSearch(
        this.client,
        itemName,
        this.purchaseHistory,
        (results) => this.selectBestProduct(results, itemName)
      );

      if (!match) {
        console.log("❌ Not found (tried exact, history match, and fallback queries)");
        this.failedItems.push(itemName);
        return false;
      }

      await this.client.cart.addProductToCart(match.product.id, 1);
      console.log(`✅ Added (${match.matchedVia})`);
      this.addedItems.push(itemName);
      return true;
    } catch (error) {
      console.log(`❌ Error: ${error.message}`);
      this.failedItems.push(itemName);
      return false;
    }
  }

  async addAllGroceries() {
    console.log(`📦 Adding ${GROCERY_LIST.length} items to cart...\n`);

    for (const item of GROCERY_LIST) {
      await this.searchAndAddItem(item);
      // Small delay between requests to avoid rate limiting
      await new Promise((resolve) => setTimeout(resolve, 500));
    }
  }

  showSummary() {
    console.log("\n" + "=".repeat(50));
    console.log(`✅ Added: ${this.addedItems.length} items`);

    if (this.failedItems.length > 0) {
      console.log(`❌ Failed: ${this.failedItems.length} items`);
      console.log("\nFailed items (add manually if needed):");
      this.failedItems.forEach((item) => console.log(`  - ${item}`));
    }

    console.log("\n📝 Next steps:");
    console.log("  1. Open Picnic app");
    console.log("  2. Review your cart");
    console.log("  3. Select delivery slot and checkout");
    console.log("\n✨ Automation complete!");
  }
}

function askQuestion(question) {
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
  });

  return new Promise((resolve) => {
    rl.question(question, (answer) => {
      rl.close();
      resolve(answer);
    });
  });
}

async function loadGroceryListFromServer(serverUrl = "https://192.168.178.78:3000") {
  try {
    const response = await fetch(`${serverUrl}/api/picnic-sync`, {
      method: "POST",
      headers: { "Content-Type": "application/json" }
    });

    if (!response.ok) throw new Error(`Server error: ${response.status}`);

    const data = await response.json();
    if (data.items && Array.isArray(data.items)) {
      GROCERY_LIST = data.items;
      console.log(`✅ Loaded ${GROCERY_LIST.length} items from meal hub\n`);
      return true;
    }
  } catch (error) {
    console.log("⚠️  Could not load from server, using command-line argument\n");
    return false;
  }
  return false;
}

async function runAutomation(username, password) {
  try {
    const automation = new PicnicAutomation(username, password);

    if (!(await automation.login())) {
      return { success: false, error: "Login failed" };
    }

    await automation.loadPurchaseHistory();
    await automation.clearCart();
    await automation.addAllGroceries();

    return {
      success: true,
      addedItems: automation.addedItems,
      failedItems: automation.failedItems,
      summary: {
        added: automation.addedItems.length,
        failed: automation.failedItems.length
      }
    };
  } catch (error) {
    return { success: false, error: error.message };
  }
}

async function main() {
  try {
    let username = process.env.PICNIC_USERNAME;
    let password = process.env.PICNIC_PASSWORD;

    // Try to load shopping list from server
    await loadGroceryListFromServer();

    if (!username) {
      username = await askQuestion("Enter your Picnic email: ");
    }
    if (!password) {
      password = await askQuestion("Enter your Picnic password: ");
    }

    const result = await runAutomation(username, password);

    if (result.success) {
      console.log("\n" + "=".repeat(50));
      console.log(`✅ Added: ${result.summary.added} items`);
      if (result.summary.failed > 0) {
        console.log(`❌ Failed: ${result.summary.failed} items`);
      }
      console.log("\n📝 Next steps:");
      console.log("  1. Open Picnic app");
      console.log("  2. Review your cart");
      console.log("  3. Select delivery slot and checkout");
      console.log("\n✨ Automation complete!");
    } else {
      console.error("❌ Error:", result.error);
      process.exit(1);
    }
  } catch (error) {
    console.error("❌ Error:", error.message);
    process.exit(1);
  }
}

// Allow this to be required as a module (for server integration)
module.exports = { PicnicAutomation, runAutomation, loadGroceryListFromServer };

// Run if called directly from CLI
if (require.main === module) {
  main();
}
